pkg/engine/task/pod_exec.go

```diff
@@ -0,0 +1,232 @@
+	if err := <-errCh; err != nil {
+		return 0, fmt.Errorf("failed to get the size of %s, err: %v, %s", file, err, raw)
+	}
+
+	raw = raw[:len(raw)-2] // remove trailing \n\r
+	size, err := strconv.ParseInt(raw, 10, 64)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse the size '%s' of the file %s : %v", raw, file, err)
+	}
+
+	return size, nil
+}
+
+// DownloadFile fetches a file from a remote pod. It runs `tar cf - file` command and streams contents
+// of the file via the stdout. Locally, the tar file is extracted into the passed afero filesystem where
+// it is saved under the same path. Afero filesystem is used to allow the caller downloading and persisting
+// of multiple files concurrently (afero filesystem is thread-safe).
+func DownloadFile(fs afero.Fs, file string, pod *v1.Pod, restCfg *rest.Config) error {
+	readout, writeout := io.Pipe()
+	stderr := bytes.Buffer{}
+
+	pe := PodExec{
+		RestCfg:       restCfg,
+		PodName:       pod.Name,
+		PodNamespace:  pod.Namespace,
+		ContainerName: pipePodContainerName,
+		Args:          []string{"tar", "cf", "-", file},
+		In:            nil,
+		Out:           writeout,
+		Err:           &stderr,
+	}
+
+	// IMPORTANT:
+	// When downloading files using remotecommand.Executor, we pump the contents of the tar file through
+	// the stdout. PodExec.Run() calls the stream executor, which first writes stdout and stderr of the
+	// command and then returns with an exit code. Since we're using io.Pipe reader and writer which is
+	// tread-safe but SYNCHRONOUS, Run() call will not return until we've consumed the streams (and will
+	// block the execution). For more details on how the underlying streams are copied see remotecommand/v4.go:54
+	//
+	// TL;DR:
+	//  - execute PodExec.Run() in a goroutine when using io.Pipe for Out or Err streams, they
+	//    have to be consumed first because io.Pipe is synchronous (and thread-safe)
+	//  - there seems to be a bug when consuming the error stream of the above command: EOF is never
+	//    sent so reading it never ends ¯\_(ツ)_/¯ This is why we use simple bytes.Buffer for the
+	//    Err stream (it is only used in the error message)
+	//
+	// See `kubectl cp` copyFromPod method for another example:
+	// https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/cp/cp.go#L291
+	errCh := make(chan error, 1)
+	go func() {
+		defer writeout.Close() // Close never returns an error
+		errCh <- pe.Run()
+	}()
+
+	if err := untarFile(fs, readout, file); err != nil {
+		return fmt.Errorf("failed to untar pipe file: %v", err)
+	}
+
+	// THIS CHECK HAS TO HAPPEN AFTER WE CONSUME THE STDOUT/STDERR READER.
+	if err := <-errCh; err != nil {
+		return fmt.Errorf("failed to copy pipe file. err: %v, stderr: %s", err, stderr.String())
+	}
+
+	return nil
+}
+
+// untarFile extracts a tar file from the passed reader using passed file name.
+func untarFile(fs afero.Fs, r io.Reader, fileName string) error {
```

Comment:
did you harass me months ago that we should abstract tarring / untarring :) ?

Reply:
I certainly didn't "haras" anybody 👅 And the fact that we have this tar-iterator 3x times in our code-base is lamentable. I started to unify them and it turned out to be cumbersome because of what this method should return (parsed objects, files in `fs` etc) so I decided to go a different path and make this method very specific to the use case (e.g. it can only untar files, no folders).
I intend to come back to this issue later.